spark3/src/main/java/org/apache/iceberg/actions/Spark3MigrateAction.java

```diff
@@ -0,0 +1,155 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.SnapshotSummary;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.spark.SparkSessionCatalog;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagedTable;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Some;
+import scala.collection.JavaConverters;
+
+/**
+ * Takes a Spark table in the sourceCatalog and attempts to transform it into an Iceberg
+ * Table in the same location with the same identifier. Once complete the identifier which
+ * previously referred to a non-iceberg table will refer to the newly migrated iceberg
+ * table.
+ */
+class Spark3MigrateAction extends Spark3CreateAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3MigrateAction.class);
+  private static final String BACKUP_SUFFIX = "_BACKUP_";
+
+  Spark3MigrateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableName) {
+    super(spark, sourceCatalog, sourceTableName, sourceCatalog, sourceTableName);
+  }
+
+  @Override
+  public Long execute() {
+    Table icebergTable;
+
+    // Move source table to a new name, halting all modifications and allowing us to stage
+    // the creation of a new Iceberg table in its place
+    String backupName = sourceTableName().name() + BACKUP_SUFFIX;
+    Identifier backupIdentifier = Identifier.of(sourceTableName().namespace(), backupName);
+    try {
+      destCatalog().renameTable(sourceTableName(), backupIdentifier);
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException("Cannot find table to migrate", e);
+    } catch (TableAlreadyExistsException e) {
+      throw new IllegalArgumentException("Cannot rename migration source to backup name", e);
+    }
+
+    StagedTable stagedTable = null;
+    boolean threw = true;
+    try {
+      Map<String, String> newTableProperties = new HashMap<>();
+      newTableProperties.put(TableCatalog.PROP_PROVIDER, "iceberg");
+      newTableProperties.putAll(JavaConverters.mapAsJavaMapConverter(v1SourceTable().properties()).asJava());
+      newTableProperties.putAll(tableLocationProperties(sourceTableLocation()));
+      newTableProperties.putAll(additionalProperties());
+
+      stagedTable = destCatalog().stageCreate(Identifier.of(
```

Comment:
It looks like the staging part is same in both. Can we factor this out into the parent class?

```
  protected StagedTable stageDestTable() {
    try {
      Map<String, String> props = targetTableProps();
      StructType schema = sourceTable.schema();
      Transform[] partitioning = sourceTable.partitioning();
      return destCatalog.stageCreate(destTableIdent, schema, partitioning, props);
    } catch (NoSuchNamespaceException e) {
      throw new IllegalArgumentException("Cannot create a new table in a namespace which does not exist", e);
    } catch (TableAlreadyExistsException e) {
      throw new IllegalArgumentException("Destination table already exists", e);
    }
  }

  protected abstract Map<String, String> targetTableProps();

```

Then we can override `targetTableProps` in each subclass and make sure we add extra metadata like `migrated` or `snapshot` flags if we want to.

Reply:
Sounds good to me